src/OpenDebugAD7/TracepointManager.cs

```diff
@@ -0,0 +1,229 @@
+        {
+            m_breakpointLogMessages = new Dictionary<uint, string>();
+
+            // Matches strings that are in { } 
+            // or upper case strings that begin with $.
+            regex = new Regex("\\$[A-Z]+|{.*?}", RegexOptions.Compiled);
+
+            propertyInfoFlags = enum_DEBUGPROP_INFO_FLAGS.DEBUGPROP_INFO_NAME |
+                enum_DEBUGPROP_INFO_FLAGS.DEBUGPROP_INFO_VALUE |
+                enum_DEBUGPROP_INFO_FLAGS.DEBUGPROP_INFO_TYPE |
+                enum_DEBUGPROP_INFO_FLAGS.DEBUGPROP_INFO_ATTRIB |
+                enum_DEBUGPROP_INFO_FLAGS.DEBUGPROP_INFO_PROP |
+                enum_DEBUGPROP_INFO_FLAGS.DEBUGPROP_INFO_FULLNAME |
+                (enum_DEBUGPROP_INFO_FLAGS)enum_DEBUGPROP_INFO_FLAGS110.DEBUGPROP110_INFO_FORCE_REAL_FUNCEVAL |
+                (enum_DEBUGPROP_INFO_FLAGS)enum_DEBUGPROP_INFO_FLAGS110.DEBUGPROP110_INFO_NOSIDEEFFECTS;
+
+            flags = enum_EVALFLAGS.EVAL_RETURNVALUE |
+                enum_EVALFLAGS.EVAL_NOEVENTS |
+                (enum_EVALFLAGS)enum_EVALFLAGS110.EVAL110_FORCE_REAL_FUNCEVAL |
+                enum_EVALFLAGS.EVAL_NOSIDEEFFECTS;
+
+        }
+
+        internal void Add(uint breakpointId, string logMessage)
+        {
+            lock (_lock)
+            {
+                m_breakpointLogMessages.Add(breakpointId, logMessage);
+            }
+        }
+
+        internal void Remove(uint breakpointId)
+        {
+            lock (_lock)
+            {
+                m_breakpointLogMessages.Remove(breakpointId);
+            }
+        }
+
+        internal void Replace(uint breakpointId, string newLogMesssage)
+        {
+            lock (_lock)
+            {
+                m_breakpointLogMessages[breakpointId] = newLogMesssage;
+            }
+        }
+
+        internal bool Contains(uint breakpointId)
+        {
+            lock (_lock)
+            {
+                return m_breakpointLogMessages.ContainsKey(breakpointId);
+            }
+        }
+
+        internal bool TryGetValue(uint breakpointId, out string logMessage)
+        {
+            lock (_lock)
+            {
+                return m_breakpointLogMessages.TryGetValue(breakpointId, out logMessage);
+            }
+        }
+
+        internal string GetLogMessage(uint breakpointId, IDebugThread2 pThread, uint radix)
+        {
+            string logMessage;
+            if (!this.TryGetValue(breakpointId, out logMessage))
+            {
+                return string.Empty;
+            }
+
+            // There is strings to interpolate in the log message.
+            if (regex.IsMatch(logMessage))
+            {
+                logMessage = GetInterpolatedLogMessage(logMessage, pThread, radix);
+            }
+
+            return logMessage;
+        }
+
+        private string GetInterpolatedLogMessage(string logMessage, IDebugThread2 pThread, uint radix)
+        {
+            if (pThread == null)
+            {
+                return logMessage;
```

Comment:
>logMessage [](start = 23, length = 10)

Shouldn't these cases be returning an error rather than the log message?

Reply:
Not sure what error to say, `unable to evaluate logMessage`?

Reply:
Yup